mqtt-protocol/src/jsMain/kotlin/it/unibo/pulvreakt/mqtt/MqttProtocol.kt

```diff
@@ -1,47 +1,157 @@
+
+    override lateinit var di: DI
+    private val logger = KotlinLogging.logger("MqttProtocol")
 
-    override suspend fun writeToChannel(from: Entity, to: Entity, message: ByteArray): Either<ProtocolError, Unit> {
-        TODO("Not yet implemented")
+    private val registeredTopics = mutableMapOf<Pair<Entity, Entity>, String>()
+    private val topicChannels = mutableMapOf<String, MutableSharedFlow<ByteArray>>()
+    private lateinit var client: MqttClient
+
+    override suspend fun setupChannel(
+        source: Entity,
+        destination: Entity
+    ) {
+        logger.debug { "Setting up channel for entity $source to $destination" }
+        registeredTopics += (source to destination) to toTopics(source, destination)
+        registeredTopics += (destination to source) to toTopics(destination, source)
+        topicChannels += toTopics(source, destination) to MutableSharedFlow(1)
+        topicChannels += toTopics(destination, source) to MutableSharedFlow(1)
     }
 
-    override fun readFromChannel(from: Entity, to: Entity): Either<ProtocolError, Flow<ByteArray>> {
-        TODO("Not yet implemented")
+    override suspend fun writeToChannel(
+        from: Entity,
+        to: Entity,
+        message: ByteArray
+    ): Either<ProtocolError, Unit> = either {
+        val topic = registeredTopics[Pair(from, to)]
+        logger.debug { "Writing message $message to topic $topic" }
+
+        ensureNotNull(topic) { ProtocolError.EntityNotRegistered(to) }
+
+        val writeOptions = js("{}")
+        writeOptions.qos = qos
+        writeOptions.retain = retain
+        writeOptions.keepalive = serverKeepAlive
+
+        Either.catch {
+            client.publish(
+                topic,
+                message.decodeToString(),
+                options = writeOptions
+            )
+        }.mapLeft { ProtocolError.ProtocolException(it) }
     }
 
-    override suspend fun initialize(): Either<ProtocolError, Unit> {
-        TODO("Not yet implemented")
+    override fun readFromChannel(
+        from: Entity,
+        to: Entity
+    ): Either<ProtocolError, Flow<ByteArray>> = either {
+        val candidateTopic = ensureNotNull(registeredTopics[Pair(from, to)]) {
+            ProtocolError.EntityNotRegistered(from)
+        }
+        val channel = ensureNotNull(topicChannels[candidateTopic]) {
+            ProtocolError.EntityNotRegistered(from)
+        }
+        logger.debug { "Reading from topic $candidateTopic" }
+        channel.asSharedFlow()
     }
 
-    override suspend fun finalize(): Either<ProtocolError, Unit> {
-        TODO("Not yet implemented")
+    override suspend fun initialize(): Either<ProtocolError, Unit> = either {
+        Either.catch {
+            logger.debug { "entering init" }
+            val connectOptions = js(
+                "{" +
+                    "  protocolId: 'MQTT'," +
+                    "  protocolVersion: 5," +
+                    "  clean: false," +
+                    "}"
```

Comment:
Use a multiline string, it is more clear.
```suggestion
                """
                    {
                        protocolId: 'MQTT',
                        protocolVersion: 5,
                        clean: false,"
                    }
                """.trimIndent()
```

Reply:
sorry, but when i tried your suggestion the code generated a 'Argument must be a string constant' error so i think it's better to just keep it this way even tho it's slightly less clear